fabric-recipe-api-v1/src/main/java/net/fabricmc/fabric/mixin/recipe/ShapedRecipeSerializerMixin.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.mixin.recipe;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+import it.unimi.dsi.fastutil.objects.Object2CharMap;
+import it.unimi.dsi.fastutil.objects.Object2CharOpenHashMap;
+import org.spongepowered.asm.mixin.Mixin;
+
+import net.minecraft.recipe.Ingredient;
+import net.minecraft.recipe.ShapedRecipe;
+import net.minecraft.util.collection.DefaultedList;
+import net.minecraft.util.registry.Registry;
+
+import net.fabricmc.fabric.api.recipe.v1.serializer.FabricRecipeSerializer;
+
+@Mixin(ShapedRecipe.Serializer.class)
+public abstract class ShapedRecipeSerializerMixin implements FabricRecipeSerializer<ShapedRecipe> {
```

Comment:
If you really write this crazy system... I would prefer you to change the event so people can register `net.minecraft.data.server.recipe.RecipeJsonProvider` instead. That thing still has a lot of functionalities as recipes, such as `Identifier getIdentifier()` which you uses.

Reply:
Or you can create another interface that has getters for both a RecipeJsonProvider and a Recipe. These two interfaces probably cannot be implemented together due to both having a `RecipeSerializer<?> getSerializer()` that probably has different obf/intermediaries.

Reply:
I'm not sure to understand what you're asking here.